src/software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_tactic.cpp

```diff
@@ -0,0 +1,65 @@
+#include "software/ai/hl/stp/tactic/shadow_enemy/shadow_enemy_tactic.h"
+
+#include "software/ai/hl/stp/action/stop_action.h"  // TODO (#1888): remove this dependency
+
+
+ShadowEnemyTactic::ShadowEnemyTactic()
+    : Tactic(false, {RobotCapability::Move, RobotCapability::Kick}),
+      fsm(),
+      control_params{ShadowEnemyFSM::ControlParams{.enemy_threat    = std::nullopt,
+                                                   .shadow_distance = 0,
+                                                   .steal_and_chip_distance = 0}}
+{
+}
+
+void ShadowEnemyTactic::updateWorldParams(const World &world) {}
+
+void ShadowEnemyTactic::updateControlParams(std::optional<EnemyThreat> enemy_threat,
+                                            double shadow_distance,
+                                            double steal_and_chip_distance)
+{
+    control_params.enemy_threat            = enemy_threat;
+    control_params.shadow_distance         = shadow_distance;
+    control_params.steal_and_chip_distance = steal_and_chip_distance;
+}
+
+double ShadowEnemyTactic::calculateRobotCost(const Robot &robot, const World &world) const
+{
+    if (!control_params.enemy_threat)
+    {
+        return 0;
+    }
+    // Prefer robots closer to the enemy being shadowed
+    // We normalize with the total field length so that robots that are within the field
+    // have a cost less than 1
+    double cost =
+        (robot.position() - control_params.enemy_threat->robot.position()).length() /
+        world.field().totalXLength();
```

Comment:
wouldn't this be inaccurate if we're shadowing some distance away from the enemy robot?

Reply:
Hmm, that is a good point. The cost also depends on whether or not we want to block the pass, block the net, or steal and chip the ball. I'm unsure how precise we want the cost function to be. I exposed some of the functions in block_pass_or_net such as `findBlockPassPoint` and `findBlockNetPoint`, would it suffice to call one of these functions instead?

Reply:
It should be as precise as possible, but without depending on the state of the FSM, since that's too complicated. What you suggested sounds about right

Reply:
Fixed